File: 50-powx-n/powx-n.py
class Solution(object):
    def myPow(self, x, n):
        if n == 0:
            return 1.0
        
        N = n
        if N < 0:
            x = 1 / x
            N = -N
        
        result = 1.0
        prod = x
        
        while N > 0:
            if N % 2 == 1:
                result *= prod
            prod *= prod
            N //= 2
        
        return result
        

File: 50-powx-n/powx-n.java
class Solution {
    public double myPow(double x, int n) {
        if (n == 0) return 1;
        if (n == Integer.MIN_VALUE) {
            x = 1 / x;
            n = Integer.MAX_VALUE;
            return x * myPow(x, n);
        }
        if (n < 0) {
            x = 1 / x;
            n = -n;
        }
        double result = 1;
        while (n > 0) {
            if ((n % 2) == 1) result *= x;
            x *= x;
            n /= 2;
        }
        return result;
    }
}